sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/xml/parsers/StaxXmlGenerator.scala

```diff
@@ -71,9 +72,11 @@ private[xml] object StaxXmlGenerator {
           writer.writeAttribute(name.substring(options.attributePrefix.length), v.toString)
 
         // For ArrayType, we just need to write each as XML element.
-        case (ArrayType(ty, _), v: scala.collection.Seq[_]) =>
-          v.foreach { e =>
-            writeChildElement(name, ty, e)
+        case (ArrayType(ty, _), v: ArrayData) =>
+          var i = 0;
+          while (i < v.numElements()) {
```

Comment:
Nit, `(0 until v.numElements()).foreach { i =>
    writeChildElement(name, ty, v.get(i, ty))
}` to avoid `var`